src/extensions/mod.rs

```diff
@@ -130,28 +104,96 @@ impl RelmListBoxExt for gtk::ListBox {
     }
 }
 
-fn iter_children(widget: &gtk::Widget) -> impl Iterator<Item = gtk::Widget> {
-    let mut widget = widget.first_child();
+/// An iterator over container children.
+#[derive(Debug)]
+pub struct ChildrenIterator<T: RelmIterChildrenExt> {
+    start: Option<T::Child>,
+    end: Option<T::Child>,
+    done: bool,
+}
+
+impl<T: RelmIterChildrenExt> ChildrenIterator<T> {
+    /// Create a new iterator over children of `widget`.
+    pub fn new(widget: &T) -> Self {
+        let widget = widget.as_ref();
+        let start = widget.first_child().map(|child| {
+            child
+                .downcast::<T::Child>()
+                .expect("The type of children does not match.")
+        });
+        let end = widget.last_child().map(|child| {
+            child
+                .downcast::<T::Child>()
+                .expect("The type of children does not match.")
+        });
+        let done = start.is_none();
+        ChildrenIterator { start, end, done }
+    }
+}
 
-    std::iter::from_fn(move || {
-        if let Some(child) = widget.take() {
-            widget = child.next_sibling();
-            return Some(child);
+impl<T: RelmIterChildrenExt> Iterator for ChildrenIterator<T> {
+    type Item = T::Child;
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.done {
+            if let Some(start) = self.start.take() {
+                if self.start == self.end {
+                    self.done = true;
+                }
+                self.start = start.next_sibling().map(|child| {
+                    child
+                        .downcast::<T::Child>()
+                        .expect("The type of children does not match.")
+                });
+                return Some(start);
+            }
         }
 
         None
-    })
```

Comment:
If you use 
```rust
if self.done {
    None
else {
    ....
    Some(start)
}
```

You can get rid of the `return` and make the program flow more obvious. Same for `DoubleEndedIterator`.